filters.lib

```diff
@@ -2797,6 +2797,102 @@ svf = environment {
 	hs(f,q,g)   = svf(8, f, q, g);
 };
 
+//----------(fi.)SVFTPT---------------------------------------------------------
+//
+// Topology-preserving transform implementation following Zavalishin's method.
+//
+// Outputs: lowpass, highpass, bandpass, normalised bandpass, notch, allpass, 
+// peaking.
+//
+// The ouptputs can be recalled by specifying the environment; for example:
+//      SVFTPT.LP2(.707, 1000.0)
+//
+// Even though the implementation is different, the characteristics of this
+// filter are comparable to those of the "svf" environment in this library.
+//
+// #### Usage:
+//
+// _ : SVFTPT(Q, CF).xxx : _
+//
+// #### Where:
+//
+// * `Q`: resonance.
+// * `CF`: cutoff in Hz.
+//------------------------------------------------------------------------------
+declare SVFTPT author "Dario Sanfilippo";
+declare SVFTPT copyright
+    "Copyright (C) 2024 Dario Sanfilippo <sanfilippo.dario@gmail.com>";
+declare SVFTPT license "MIT License";
+SVFTPT = environment {
+    SVF(Q, CF, x) = f ~ si.bus(2) : (! , ! , _ , _ , _ , _ , _ , _ , _)
+        with {
+            g = tan(CF * ma.PI * ma.T);
+            R2 = 1.0 / Q;
+            gPlusR2 = g + R2;
+            f(s0, s1) = u0 , u1 , LP , HP , BP , BPNorm , Notch , AP , Peaking
+                with {
+                    HP = (x - s0 * gPlusR2 - s1) / (1.0 + g * gPlusR2);
+                    v0 = HP * g;
+                    BP = s0 + v0;
+                    v1 = BP * g;
+                    LP = s1 + v1;
+                    BPNorm = BP * R2;
+                    Notch = x - BPNorm;
+                    AP = x - BP * (R2 + R2);
+                    Peaking = LP - HP;
+                    u0 = v0 + BP;
+                    u1 = v1 + LP;
+                };
+        };
+    LP2(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 0);
+    HP2(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 1);
+    BP2(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 2);
+    BP2Norm(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 3);
+    Notch2(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 4);
+    AP2(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 5);
+    Peaking2(Q, CF, x) = SVF(Q, CF, x) : ba.selectn(7, 6);
+};
+
+//----------(fi.)dynamicSmoother------------------------------------------------
+//
+// Adaptive smoother based on Andy Simper's paper:
+//      https://cytomic.com/files/dsp/DynamicSmoothing.pdf
+//
+// This filter uses both the lowpass and bandpass outputs of a 
+// state-variable filter. The lowpass is used to smooth out the input signal,
+// the bandpass, which is a smoothed out version of the highpass, provides
+// information on the rate of change of the input. Hence, the bandpass signal
+// can be used to adjust the cutoff of the filter to quickly follow the input's
+// fast and large variations while effectively filtering out local 
+// perturbations.
+//
+// This implementation does not use an approximation for the CF computation,
```

Comment:
Thanks for this wonderful filter!
What do you mean by this?

Reply:
This implementation just changes the CF of the SVF based on the BP output. The SVF computes the gain using tan(), Andy's paper suggests two approximations for the g-coeff.

As I said the other day, I wasn't able to test why the implementations in the paper fail for very low base CFs. One thing to do would be to implement the clipping for the computation of "g" as I did and see if that alone makes it better with the approximations.

Reply:
> As I said the other day, I wasn't able to test why the implementations in the paper fail for very low base CFs.

Where was that? I didn't find it on the Discord.

What I *did* seen [on Discord}(https://discord.com/channels/1078607704769179729/1078607704769179732/1305876664584241172) was this:
> I'd first check the impulse response of Andy's filters with the SVF with the Q at .5, but Andy is very competent so I'd tend to trust what he says on the paper. Then, I'd use the exact same mapping as I did without approximation.

I was hoping that you meant to implement the version in the paper, with Andy's mapping.
But maybe I misread you, due to wishful thinking no doubt!  :sweat_smile: